### PR TITLE
Metrics: Add MetricDescriptor.

### PR DESCRIPTION
**Pull request by songy23**
[Proto definition](https://github.com/census-instrumentation/opencensus-proto/blob/master/opencensus/proto/stats/metrics/metrics.proto#L54-L96):
```proto
// Defines a metric type and its schema.
message MetricDescriptor {
  // The metric type, including its DNS name prefix. It must be unique.
  string name = 1;

  // A detailed description of the metric, which can be used in documentation.
  string description = 2;

  // The unit in which the metric value is reported. Follows the format
  // described by http://unitsofmeasure.org/ucum.html.
  string unit = 3;

  // The kind of metric. It describes how the data is reported.
  //
  // A gauge is an instantaneous measurement of a value.
  //
  // A cumulative measurement is a value accumulated over a time interval. In
  // a time series, cumulative measurements should have the same start time and
  // increasing end times, until an event resets the cumulative value to zero
  // and sets a new start time for the following points.
  enum Type {
    // Do not use this default value.
    UNSPECIFIED = 0;

    // Integer gauge.
    GAUGE_INT64 = 1;

    // Floating point gauge.
    GAUGE_DOUBLE = 2;

    // Integer cumulative measurement.
    CUMULATIVE_INT64 = 3;

    // Floating point cumulative measurement.
    CUMULATIVE_DOUBLE = 4;

    // Distribution cumulative measurement.
    CUMULATIVE_DISTRIBUTION = 5;
  }
  Type type = 4;

  // The label keys associated with the metric descriptor.
  repeated LabelKey label_keys = 5;
}
```

### CODE REVIEW

File: metrics/src/main/java/io/opencensus/metrics/MetricDescriptor.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+   *
+   * @since 0.16
+   */
+  public enum Type {
+    /**
+     * Unspecified type. Do not use this as default value.
+     *
+     * @since 0.16
+     */
+    UNSPECIFIED,
```
Comment (sebright) [inline metrics/src/main/java/io/opencensus/metrics/MetricDescriptor.java]: Do we need UNSPECIFIED in the Java implementation?  When would it be used?

File: metrics/src/test/java/io/opencensus/metrics/MetricDescriptorTest.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    thrown.expect(NullPointerException.class);
+    thrown.expectMessage("labelKeys");
+    MetricDescriptor.create(METRIC_NAME_1, DESCRIPTION, UNIT, Type.GAUGE_DOUBLE, null);
+  }
+
+  @Test
+  public void preventNullLabelKey() {
+    List<LabelKey> keys = Arrays.asList(KEY_1, null);
+    thrown.expect(NullPointerException.class);
+    thrown.expectMessage("labelKey");
```
Comment (sebright) [inline metrics/src/test/java/io/opencensus/metrics/MetricDescriptorTest.java]: Suggestion: It might be better to test the exact error message, since this one is a substring of the error message from the test above.  For example, you could use `thrown.expectMessage(CoreMatchers.equalTo("labelKey"))`.

File: metrics/src/main/java/io/opencensus/metrics/MetricDescriptor.java (hunk 1 above)
Comment (songy23) [reply metrics/src/main/java/io/opencensus/metrics/MetricDescriptor.java]: Short answer is I don't know.. I didn't see any specification on `Type.UNSPECIFIED` in proto or specs. I think we can ignore this value for now and add it later if needed.

File: metrics/src/test/java/io/opencensus/metrics/MetricDescriptorTest.java (hunk 2 above)
Comment (songy23) [reply metrics/src/test/java/io/opencensus/metrics/MetricDescriptorTest.java]: That's true - updated. Same for other checks on null list element.

File: metrics/src/test/java/io/opencensus/metrics/TimeSeriesTest.java [hunk 3]
```diff
@@ -27,6 +27,7 @@
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import org.hamcrest.CoreMatchers;
```
Comment (bogdandrutu) [inline metrics/src/test/java/io/opencensus/metrics/TimeSeriesTest.java]: Why is this needed?

Comment (songy23) [reply metrics/src/test/java/io/opencensus/metrics/TimeSeriesTest.java]: https://github.com/census-instrumentation/opencensus-java/pull/1291#discussion_r201159089:
> Suggestion: It might be better to test the exact error message, since this one is a substring of the error message from the test above. For example, you could use thrown.expectMessage(CoreMatchers.equalTo("labelKey")).

For example there's one exception with message "labelValue" and one with "labelValues",  CoreMatchers can be used to differentiate them.
